File: lib/CSharp/FSMContext.cs
//
// The contents of this file are subject to the Mozilla Public
// License Version 1.1 (the "License"); you may not use this file
// except in compliance with the License. You may obtain a copy
// of the License at http://www.mozilla.org/MPL/
// 
// Software distributed under the License is distributed on an
// "AS IS" basis, WITHOUT WARRANTY OF ANY KIND, either express or
// implied. See the License for the specific language governing
// rights and limitations under the License.
// 
// The Original Code is  State Machine Compiler(SMC).
// 
// The Initial Developer of the Original Code is Charles W. Rapp.
// Portions created by Charles W. Rapp are
// Copyright (C) 2004. Charles W. Rapp.
// All Rights Reserved.
// 
// Contributor(s):
//   Eitan Suez contributed examples/Ant.
//   (Name withheld) contributed the C# code generation and
//   examples/C#.
//
// RCS ID
// $Id$
//
// statemap.java --
//
//  This package defines the FSMContext class which is inherited
//  by the smc-generated application FSM context class.
//
// CHANGE LOG
// (See the bottom of this file.)
//

using System;
using System.IO;
using System.Diagnostics;

namespace statemap
{
    // State change event. Contains the previous and new state.
    public class StateChangeEventArgs :
        EventArgs
    {
    //-----------------------------------------------------------
    // Member functions.
    //

        // Constructor.
        public StateChangeEventArgs(string fsmName,
                                    string transitionType,
                                    State previousState,
                                    State newState)
        {
            _fsmName = fsmName;
            _transitionType = transitionType;
            _previousState = previousState;
            _newState = newState;
        } // end of StateChangeEventArgs(string, State, State)

        public string FSMName()
        {
            return (_fsmName);
        } // end of FSMName()

        public string TransitionType()
        {
            return (_transitionType);
        } // end of TransitionType()

        public State PreviousState()
        {
            return (_previousState);
        } // end of PreviousState()

        public State NewState()
        {
            return (_newState);
        } // end of NewState()

    //-----------------------------------------------------------
    // Member data.
    //

        private readonly string _fsmName;
        private readonly string _transitionType;
        private readonly State _previousState;
        private readonly State _newState;
    } // end of class StateChangeEventArgs

    // Delegate declaration.
    public delegate void StateChangeEventHandler(
         object sender, StateChangeEventArgs args);

    // statemap.FSMContext --
    //
    //  Base class for the smc-generated application FSM
    //  context class.

    [Serializable]
    public abstract class FSMContext
    {
    //-----------------------------------------------------------
    // Member functions.
    //

        public FSMContext()
        {
            // There is no state until the application explicitly
            // sets the initial state.
            _name = "FSMContext";
            _state = null;
            _transition = "";
            _previousState = null;
            _stateStack = null;
            _debugFlag = false;
            _debugStream = null;
        } // end of FSMContext()

        // The state change event.
        public event StateChangeEventHandler StateChange;

        // The finite state machine name property.
        public string Name
        {
            get
            {
                return (_name);
            }
            set
            {
                _name = value;
            }
        } // end of Name()

        // DEPRECATED
        // As of v. 4.3.3, System.Diagnostics.Trace is
        // used instead of the _debugFlag, _debugStream
        // pair.
        // Used to enable debugging output
        public bool Debug
        {
            get
            {
                return _debugFlag;
            }
            set
            {
                _debugFlag = value;
            }
        }

        // DEPRECATED
        // As of v. 4.3.3, System.Diagnostics.Trace is
        // used instead of the _debugFlag, _debugStream
        // pair. _debugStream will always be null.
        // Used to set the output text writer.
        public TextWriter DebugStream
        {
            get
            {
                return _debugStream;
            }
            set
            {
                // DEPRECATED.
                // _debugStream = value;
            }
        }

        // Is this state machine in a transition? If state is
        // null, then true; otherwise, false.
        public bool InTransition
        {
            get 
            {
                return(_state == null ? true : false);
            }
        }

        public void SetState(State state)
        {
            StateChangeEventArgs e =
                new StateChangeEventArgs(
                    _name, "SET", _state, state);

#if TRACE
            Trace.WriteLine("NEW STATE    : " +    state.Name);
#endif

            _state = state;

            OnStateChange(e);

            return;
        }

        public void ClearState()
        {
            _previousState = _state;
            _state = null;

            return;
        }

        public State PreviousState
        {
            get
            {
                if (_previousState != null)
                {
                    return(_previousState);
                }

                throw
                    new System.NullReferenceException(
                        "Previous state not set.");
            }
        }

        public void PushState(State state)
        {
            StateChangeEventArgs e =
                new StateChangeEventArgs(
                    _name, "PUSH", _state, state);

#if TRACE
            Trace.WriteLine("PUSH TO STATE: " +    state.Name);
#endif

            if (_state != null)
            {
                if (_stateStack == null)
                {
                    _stateStack = new System.Collections.Stack();
                }

                _stateStack.Push(_state);
            }

            _state = state;

            OnStateChange(e);

            return;
        }

        public void PopState()
        {
            if (_stateStack.Count == 0)
            {
#if TRACE
                Trace.WriteLine("POPPING ON EMPTY STATE STACK.");
#endif

                throw new
                    System.InvalidOperationException(
                        "popping an empty state stack");
            }
            else
            {
                State nextState = (State) _stateStack.Pop();
                StateChangeEventArgs e =
                    new StateChangeEventArgs(
                        _name, "POP", _state, nextState);

                // The pop method removes the top element
                // from the stack and returns it.
                _state = nextState;

#if TRACE
                Trace.WriteLine("POP TO STATE : " + _state.Name);
#endif

                OnStateChange(e);
            }

            return;
        }

        public void EmptyStateStack()
        {
            _stateStack.Clear();
        }

        public string GetTransition()
        {
            return _transition;
        }

        // Release all acquired resources.
        ~FSMContext()  //TODO: Add disposable
        {
            _name = null;
            _state = null;
            _transition = null;
            _previousState = null;
            _stateStack = null;
        }

        protected virtual void OnStateChange(StateChangeEventArgs e)
        {
            if (StateChange != null)
            {
                StateChange(this, e);
            }

            return;;
        } // end of OnStateChange(StateChangeEventArgs)

    //-----------------------------------------------------------
    // Member data
    //

        // The finite state machine's unique name.
        [NonSerialized]
        protected string _name;

        // The current state.
        [NonSerialized]
        protected State _state;

        // The current transition *name*. Used for debugging
        // purposes.
        [NonSerialized]
        protected string _transition;

        // Remember what state a transition left.
        // Do no persist the previous state because an FSM should
        // be serialized while in transition.
        [NonSerialized]
        protected State _previousState;

        // This stack is used when a push transition is taken.
        [NonSerialized]
        protected System.Collections.Stack _stateStack;

        // DEPRECATED
        // As of v. 4.3.3, System.Diagnostics.Trace is
        // used instead of the _debugFlag, _debugStream
        // pair. _debugStream will always be null.
        // When this flag is set to true, this class will print
        // out debug messages.
        [NonSerialized]
        protected bool _debugFlag;

        // Write debug output to this stream.
        [NonSerialized]
        protected TextWriter _debugStream;
    } // end of class FSMContext
} // end of namespace statemap

//
// CHANGE LOG
// $Log$
// Revision 1.5  2008/01/14 19:59:23  cwrapp
// Release 5.0.2 check-in.
//
// Revision 1.4  2006/09/16 15:04:28  cwrapp
// Initial v. 4.3.3 check-in.
//
// Revision 1.3  2006/06/03 19:39:25  cwrapp
// Final v. 4.3.1 check in.
//
// Revision 1.2  2006/04/22 12:45:25  cwrapp
// Version 4.3.1
//
// Revision 1.1  2005/05/28 18:44:13  cwrapp
// Updated C++, Java and Tcl libraries, added CSharp, Python and VB.
//
// Revision 1.0  2004/09/06 16:32:15  charlesr
// Initial revision
//

File: lib/Java/FSMContext.java
//
// The contents of this file are subject to the Mozilla Public
// License Version 1.1 (the "License"); you may not use this file
// except in compliance with the License. You may obtain a copy
// of the License at http://www.mozilla.org/MPL/
// 
// Software distributed under the License is distributed on an
// "AS IS" basis, WITHOUT WARRANTY OF ANY KIND, either express or
// implied. See the License for the specific language governing
// rights and limitations under the License.
// 
// The Original Code is  State Machine Compiler(SMC).
// 
// The Initial Developer of the Original Code is Charles W. Rapp.
// Portions created by Charles W. Rapp are
// Copyright (C) 2000 - 2007. Charles W. Rapp.
// All Rights Reserved.
// 
// Contributor(s): 
//
// statemap.java --
//
//  This package defines the FSMContext class which must be
//  inherited by any Java class wanting to use an smc-generated
//  state machine.
//
// RCS ID
// $Id$
//
// CHANGE LOG
// (See the bottom of this file.)
//

package statemap;

import java.beans.PropertyChangeListener;
import java.beans.PropertyChangeSupport;
import java.io.PrintStream;
import java.io.Serializable;
import java.util.EmptyStackException;

//
// statemap.FSMContext --
//
//  All Java classes wanting to use an SMC-generated state
//  machine must extend this class. Since FSMContext provides
//  functionality, it was not possible to implement it as an
//  interface. See the SMC FAQ for how a class can use a state
//  machine when that class is already extending another class.

public abstract class FSMContext
    implements Serializable
{
//---------------------------------------------------------------
// Member functions
//

    //-----------------------------------------------------------
    // Constructors.
    //

    public FSMContext()
    {
        // There is no state until the application explicitly
        // sets the initial state.
        _state = null;
        _transition = "";
        _previousState = null;
        _stateStack = null;
        _debugFlag = false;
        _debugStream = System.err;
        _listeners = new PropertyChangeSupport(this);
    } // end of FSMContext()

    //
    // end of Constructors.
    //-----------------------------------------------------------

    //-----------------------------------------------------------
    // Get methods.
    //

    // When debug is set to true, the state machine
    // will print messages to the console.
    public boolean getDebugFlag()
    {
        return(_debugFlag && _debugStream != null);
    } // end of getDebugFlag()

    // Write the debug output to this stream.
    public PrintStream getDebugStream()
    {
        return (_debugStream == null ?
                System.err :
                _debugStream);
    } // end of getDebugStream()

    // Is this state machine in a transition? If state is null,
    // then true; otherwise, false.
    public boolean isInTransition()
    {
        return(_state == null ? true : false);
    } // end of isInTransition()

    // NOTE: getState() is defined in the SMC-generated
    // FSMContext subclass.

    public State getPreviousState()
        throws NullPointerException
    {
        if (_previousState == null)
        {
            throw (new NullPointerException());
        }
        else
        {
            return(_previousState);
        }
    } // end of getPreviousState()

    public String getTransition()
    {
        return(_transition);
    } // end of getTransition()

    //
    // end of Get methods.
    //-----------------------------------------------------------

    //-----------------------------------------------------------
    // Set methods.
    //

    public void setDebugFlag(boolean flag)
    {
        _debugFlag = flag;
        return;
    } // end of setDebugFlag(boolean)

    public void setDebugStream(PrintStream stream)
    {
        _debugStream = stream;
        return;
    } // end of setDebugStream(PrintStream)

    public void setState(State state)
    {
        State previousState = _state;

        if (getDebugFlag() == true)
        {
            getDebugStream().println("NEW STATE    : " +
                                     state.getName());
        }

        _state = state;

        // Inform any and all listeners about this state
        // change.
        _listeners.firePropertyChange(
            "State", previousState, _state);

        return;
    } // end of setState(State)

    public void clearState()
    {
        _previousState = _state;
        _state = null;

        return;
    } // end of clearState()

    public void pushState(State state)
    {
        if (_state == null)
        {
            throw (new NullPointerException());
        }

        if (getDebugFlag() == true)
        {
            getDebugStream().println("PUSH TO STATE: " +
                                     state.getName());
        }

        if (_stateStack == null)
        {
            _stateStack = new java.util.Stack<State>();
        }

        _stateStack.push(_state);
        _state = state;

        return;
    } // end of pushState(State)

    public void popState()
        throws EmptyStackException
    {
        if (_stateStack == null ||
            _stateStack.isEmpty() == true)
        {
            if (getDebugFlag() == true)
            {
                getDebugStream().println(
                    "POPPING ON EMPTY STATE STACK.");
            }

            throw (new EmptyStackException());
        }
        else
        {
            State previousState = _state;

            // The pop method removes the top element
            // from the stack and returns it.
            _state = _stateStack.pop();

            if (_stateStack.isEmpty() == true)
            {
                _stateStack = null;
            }

            if (getDebugFlag() == true)
            {
                getDebugStream().println("POP TO STATE : " +
                                      _state.getName());
            }

            // Inform any and all listeners about this state
            // change.
            _listeners.firePropertyChange(
                "State", previousState, _state);
        }

        return;
    } // end of popState()

    public void emptyStateStack()
    {
        _stateStack.clear();
        _stateStack = null;

        return;
    } // end of emptyStateStack()

    //
    // end of Set methods.
    //-----------------------------------------------------------

    // The following methods allow listeners to watch this
    // finite state machine for state changes.
    // Note: if a transition does not cause a state change,
    // then no state change event is fired.

    public void
        addStateChangeListener(PropertyChangeListener listener)
    {
        _listeners.addPropertyChangeListener("State", listener);
        return;
    }

    public void
        removeStateChangeListener(
            PropertyChangeListener listener)
    {
        _listeners.removePropertyChangeListener(
            "State", listener);
        return;
    }

//---------------------------------------------------------------
// Member data
//

    // The current state.
    transient protected State _state;

    // The current transition *name*. Used for debugging
    // purposes.
    transient protected String _transition;

    // Remember what state a transition left.
    // Do no persist the previous state because an FSM should be
    // serialized while in transition.
    transient protected State _previousState;

    // This stack is used when a push transition is taken.
    transient protected java.util.Stack<State> _stateStack;

    // When this flag is set to true, this class will print
    // out debug messages.
    transient protected boolean _debugFlag;

    // Write debug output to this stream.
    transient protected PrintStream _debugStream;

    // Store the property change listeners here.
    private PropertyChangeSupport _listeners;
} // end of class FSMContext

//
// CHANGE LOG
// $Log$
// Revision 1.9  2007/08/05 13:00:34  cwrapp
// Version 5.0.1 check-in. See net/sf/smc/CODE_README.txt for more information.
//
// Revision 1.8  2007/02/21 13:50:59  cwrapp
// Moved Java code to release 1.5.0
//
// Revision 1.7  2005/05/28 18:44:13  cwrapp
// Updated C++, Java and Tcl libraries, added CSharp, Python
// and VB.
//
// Revision 1.1  2005/02/21 19:03:38  charlesr
// Variable name clean up.
//
// Revision 1.0  2003/12/14 20:38:40  charlesr
// Initial revision
//
